Objects/unicodeobject.c

```diff
@@ -2332,6 +2332,196 @@ PyUnicodeWriter_WriteUCS4(PyUnicodeWriter *pub_writer,
 }
 
 
+static int
+unicode_export(PyObject *unicode, Py_buffer *view, uint32_t *pformat,
+               Py_ssize_t len, const void *buf,
+               int itemsize, const char *format, uint32_t internal_format)
+{
+    if (PyBuffer_FillInfo(view, unicode, (void*)buf, len,
+                          1, PyBUF_SIMPLE) < 0) {
+        return -1;
+    }
+    view->itemsize = itemsize;
+    view->format = (char*)format;
+    view->internal = (void*)(uintptr_t)internal_format;
+    *pformat = internal_format;
+    return 0;
+}
+
+
+int
+PyUnicode_Export(PyObject *unicode, uint32_t requested_formats,
+                 Py_buffer *view, uint32_t *format)
+{
+#if SIZEOF_INT == 4
+#  define BUFFER_UCS4 "I"
+#elif SIZEOF_LONG == 4
+#  define BUFFER_UCS4 "L"
+#else
+#  error "unable to find BUFFER_UCS4"
+#endif
+
+    if (!PyUnicode_Check(unicode)) {
+        PyErr_Format(PyExc_TypeError, "must be str, not %T", unicode);
+        return -1;
+    }
+    Py_ssize_t len = PyUnicode_GET_LENGTH(unicode);
+
+    // Native ASCII
+    if (PyUnicode_IS_ASCII(unicode)
+        && (requested_formats & PyUnicode_FORMAT_ASCII))
+    {
+        return unicode_export(unicode, view, format,
+                              len, PyUnicode_1BYTE_DATA(unicode),
+                              1, "B", PyUnicode_FORMAT_ASCII);
+    }
+
+    // Native UCS1
+    int kind = PyUnicode_KIND(unicode);
+    if (kind == PyUnicode_1BYTE_KIND
+        && (requested_formats & PyUnicode_FORMAT_UCS1))
+    {
+        return unicode_export(unicode, view, format,
+                              len, PyUnicode_1BYTE_DATA(unicode),
+                              1, "B", PyUnicode_FORMAT_UCS1);
+    }
+
+    // Native UCS2
+    if (kind == PyUnicode_2BYTE_KIND
+        && (requested_formats & PyUnicode_FORMAT_UCS2))
+    {
+        return unicode_export(unicode, view, format,
+                              len, PyUnicode_2BYTE_DATA(unicode),
+                              2, "H", PyUnicode_FORMAT_UCS2);
+    }
+
+    // Convert ASCII or UCS1 to UCS2
+    if (kind == PyUnicode_1BYTE_KIND
+        && requested_formats & PyUnicode_FORMAT_UCS2)
+    {
+        Py_UCS2 *ucs2 = PyMem_Malloc((len + 1) * sizeof(Py_UCS2));
+        if (!ucs2) {
+            PyErr_NoMemory();
+            return -1;
+        }
+
+        _PyUnicode_CONVERT_BYTES(Py_UCS1, Py_UCS2,
+                                 PyUnicode_1BYTE_DATA(unicode),
+                                 PyUnicode_1BYTE_DATA(unicode) + len,
+                                 ucs2);
+        ucs2[len] = 0;
+
+        return unicode_export(unicode, view, format,
```

Comment:
In the cases where there is a conversion, does the buffer need to hold a reference to the `unicode` object?  Shouldn't we instead be passing NULL here?

Reply:
I would prefer to store `unicode` as `view.obj` in all cases. So PyBuffer_Release() is able to find the `unicode_releasebuffer()` function. Technically, you're right that in UCS2 and UCS4 cases, when we allocate a buffer, we don't need to hold a reference to `unicode` to make sure that data remains valid.